sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -192,9 +207,17 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
               outputSchema,
               options.getVerifyRowValues(),
               getJarPaths(program),
-              inputGetter.getFieldAccess());
-
-      return upstream.apply(ParDo.of(calcFn)).setRowSchema(outputSchema);
+              inputGetter.getFieldAccess(),
+              errorsTransformer != null);
+
+      PCollectionTuple tuple =
+          upstream.apply(ParDo.of(calcFn).withOutputTags(rows, TupleTagList.of(errors)));
+      PCollection<BeamCalcRelError> errorPCollection =
+          tuple.get(errors).setCoder(BeamCalcRelErrorCoder.of(RowCoder.of(upstream.getSchema())));
```

Comment:
This won't work, there is no guarantee that your input row matches the input row.

If your input row is `1 as key, "a" as value`. If your SqlTransform is `SELECT CustomUdafWithError(key)`, you'd end up with a row of `1 as key` in the error. On the overall SQL transform you can end up with multiple different coders, for example if you have a join.

I think you'll actually need a coder that to encodes the coder, it's going to get complex quick. We might just want to return the string (or possibly JSON) representation of the error row instead? Can you add tests where you have errors from the same SqlTransform with different rows schemas? We are trying to move to Schema coder for everything, but I think this use case (variable row structure) isn't something we support.
cc: @reuvenlax

Reply:
sure,  I can change it to be a json string instead of row, but before I change it, I want to make sure we don't miss here something, because I prefer to send the row itself to the error transformer.

The error row is not the row *after* running the expression and getting the error, it is the origin row, as it was *before* running the DoFn, that's why I set the schema to be `upstream.getSchema()`, 
The row *after* the expression execution is a different one, and has `outputSchema`.

Also each node in the same SqlTransform create this chaining, and will assign a diff input and output schema, these schemas are per node and not per all the sql transform.
 
This is how will it look like when expanding it in DF (ignore the failed icons, this is because I abort the execution but it worked well)
![image](https://user-images.githubusercontent.com/3658085/141094730-77671b5b-1cc5-4839-8610-80eb6b379843.png)

Reply:
I added a complex test, with 2 udaf, each one produced a different error on a different schema (one before group by and one after group by)
I assert in the tests that I get 2 error rows each one with its matching schema.
from logs: (you can see that `CastUdf` and `CalculatePrice` exist in 2 diff steps.

```
INFO: SQL:
SELECT `PCOLLECTION`.`id`, `PCOLLECTION`.`country_code`, `CalculatePrice`(SUM(`CastUdf`(`PCOLLECTION`.`amount`)), `PCOLLECTION`.`currency`) AS `sum_amount`
FROM `beam`.`PCOLLECTION` AS `PCOLLECTION`
GROUP BY `PCOLLECTION`.`id`, `PCOLLECTION`.`country_code`, `PCOLLECTION`.`currency`
Nov 14, 2021 9:37:51 PM org.apache.beam.sdk.extensions.sql.impl.CalciteQueryPlanner convertToBeamRel
INFO: SQLPlan>
LogicalProject(id=[$0], country_code=[$1], sum_amount=[CalculatePrice($3, $2)])
  LogicalAggregate(group=[{0, 1, 2}], agg#0=[SUM($3)])
    LogicalProject(id=[$0], country_code=[$2], currency=[$3], $f3=[CastUdf($1)])
      BeamIOSourceRel(table=[[beam, PCOLLECTION]])
```